### PR TITLE
Release 1.2.1

### PR DESCRIPTION
**Pull request by xschildw**
Functional, needs refactor.

### CODE REVIEW

File: src/main/java/org/sagebionetworks/migration/MigrationClient.java [hunk 1]
```diff
@@ -252,37 +252,54 @@ public void migrateTypes(List<TypeToMigrateMetadata> primaryTypes, long maxBacku
...
-			long count =  dd.getCounts().getDelete();
-			if(count > 0){
-				deleteFromDestination(dd.getType(), dd.getDeleteTemp(), count, maxBackupBatchSize);
+			if (! tm.getType().equals(MigrationType.CHANGE)) {
+				migrateTypeWithRetry(salt, tm, maxBackupBatchSize, minRangeSize, timeoutMS, 3);
+			} else {
+				changeMeta = tm;
 			}
 		}
+		// Do the CHANGES
```
Comment (xschildw) [inline src/main/java/org/sagebionetworks/migration/MigrationClient.java]: If error, then don't do changes.

File: src/main/java/org/sagebionetworks/migration/MigrationClient.java [hunk 2]
```diff
@@ -252,37 +252,54 @@ public void migrateTypes(List<TypeToMigrateMetadata> primaryTypes, long maxBacku
...
 
-		// Now do all adds in the original order
-		for(int i=0; i<deltaList.size(); i++){
-			DeltaData dd = deltaList.get(i);
-			long count = dd.getCounts().getCreate();
-			if(count > 0){
-				createUpdateInDestination(dd.getType(), dd.getCreateTemp(), count, maxBackupBatchSize, timeoutMS);
+	private void migrateTypeWithRetry(String salt, TypeToMigrateMetadata tm, long maxBackupBatchSize, long minRangeSize, long timeoutMS, int maxRetry) throws Exception {
+		int numTry = 1;
+		while (numTry < maxRetry) {
```
Comment (xschildw) [inline src/main/java/org/sagebionetworks/migration/MigrationClient.java]: Move the retries down one level.

File: src/main/java/org/sagebionetworks/migration/async/AsyncMigrationRequestExecutor.java [hunk 3]
```diff
@@ -46,7 +46,11 @@ public AdminResponse execute() throws AsyncMigrationException {
 				throw new IllegalArgumentException("Response from job " + status.getJobId() + " should be AsyncMigrationResponse!");
 			}
 		} catch (TimeoutException | WorkerFailedException | SynapseException | IllegalArgumentException e) {
-			AsyncMigrationException e2 = new AsyncMigrationException("Exception in async migration job.", e);
+			String msg = "";
+			if (e.getMessage() != null) {
```
Comment (xschildw) [inline src/main/java/org/sagebionetworks/migration/async/AsyncMigrationRequestExecutor.java]: Not necessary,

File: src/test/java/org/sagebionetworks/migration/async/AsyncMigrationExecutorTest.java [hunk 4]
```diff
@@ -95,11 +99,53 @@ public void testFailed() throws Exception {
...
+			AdminResponse resp = worker.execute();
+		} catch (AsyncMigrationException e) {
+			assertTrue(e.getCause() instanceof SynapseException);
+			assertEquals(e.getMessage(), e.getCause().getMessage());
+		}
+
+		verify(mockClock, never()).sleep(anyLong());
+	}
+
+	@Test
```
Comment (xschildw) [inline src/test/java/org/sagebionetworks/migration/async/AsyncMigrationExecutorTest.java]: Not necessary.

File: src/main/java/org/sagebionetworks/migration/async/AsyncMigrationRequestExecutor.java [hunk 5]
```diff
@@ -34,24 +35,55 @@ public AsyncMigrationRequestExecutor(SynapseAdminClient client, AdminRequest req
...
+			numRetry++;
+			try {
+				jobStatus = waitForJobToComplete(jobId);
+				break;
+			} catch (TimeoutException e) {
+				throw(e);
+			} catch (WorkerFailedException e) {
+				throw(e);
+			} catch (SynapseException e) {
+				if (numRetry == maxRetry) {
```
Comment (john-hill) [inline src/main/java/org/sagebionetworks/migration/async/AsyncMigrationRequestExecutor.java:73]: greater than

File: src/main/java/org/sagebionetworks/migration/async/AsyncMigrationRequestExecutor.java [hunk 6]
```diff
@@ -34,24 +35,55 @@ public AsyncMigrationRequestExecutor(SynapseAdminClient client, AdminRequest req
...
+					logger.debug("Too many retries in waitForJobToCompleteWithRetry()");
+					throw new AsyncMigrationException("AsyncMigrationRequestExecutor: number of retries exceeded.", e);
+				}
+				try {
+					clock.sleep(1000);
+				} catch (InterruptedException e2) {
```
Comment (john-hill) [inline src/main/java/org/sagebionetworks/migration/async/AsyncMigrationRequestExecutor.java:79]: do not catch interrupts

**Pull request merged**